addons/web/controllers/json.py

```diff
@@ -244,6 +291,61 @@ def get_action_triples_():
         context.update(safe_eval(action.context, eval_context))
         return action, context, eval_context, record_id
 
+    def _web_json_2_rpc_readonly(self):
+        try:
+            model_name = request.httprequest.args['model']
+            method_name = request.httprequest.args['method']
+            Model = request.env[model_name]
+        except KeyError:
+            return False
+        for cls in Model.mro():
+            method = getattr(cls, method_name, None)
+            if method is not None and hasattr(method, '_readonly'):
+                return method._readonly
+        return False
+
+    @http.route(
+        '/json/2/<model>/<method>',
+        #methods=['POST'],  # must accept all methods, otherwise /json/<path:subpath> takes over
+        auth='bearer',
+        type='/json/2/rpc',
+        readonly=_web_json_2_rpc_readonly,
+        save_session=False
+    )
+    def web_json_2_rpc(
+        self,
+        model: str,
+        method: str,
+        ids: Sequence[int] = (),
+        kwargs: Mapping[str, Any] = frozendict(),
+        context: Mapping[str, Any] = frozendict(),
+    ):
+        if request.httprequest.method != 'POST':
+            raise MethodNotAllowed(response=request.make_json_response(
+                MethodNotAllowed.description, status=HTTPStatus.METHOD_NOT_ALLOWED))
+
+        try:
+            Model = request.env[model]
+        except KeyError as exc:
+            exc.error_response = request.make_json_response(
+                f"The model {model!r} does not exist", status=HTTPStatus.NOT_FOUND)
+            raise
+        # call_kw doesn't sanitize nor the ids, nor the context, so we don't
+        records = Model.with_context(context).browse(ids)
+        try:
+            func = get_public_method(records, method)
+        except AttributeError as exc:
+            exc.error_response = request.make_json_response(
+                exc.args[0], status=HTTPStatus.NOT_FOUND)
+            raise
+        signature = inspect.signature(func)
+        try:
+            signature.bind(records, **kwargs)
+        except TypeError as exc:
+            exc.error_response = request.make_json_response(
+                exc.args[0], status=HTTPStatus.UNPROCESSABLE_ENTITY)
+            raise
+        return func(records, **kwargs)
```

Comment:
Why not just the following? Won't the dispatcher handle the error codes?
```python
return call_kw(Model, method, [], kwargs)
```

Reply:
Does not extract context from kwargs, so not as suggested.
But we still need to handle return values (it could be as simple as handling just BaseModel return type) and fail when `@api.model` function is called with ids.

Reply:
`call_kw` extracts the `ids` and `context` from the args/kwargs, that's some very old "new v8 api" compatibility code that has never been changed. Since we create a new endpoint, I placed the ids/context *outside* of the kwargs by default, so we don't need `call_kw`'s crap. But indeed `call_kw` also does some magic with models (`return records.ids`) and we should raise an error when we call a `@api.model` decorated method with non-empty ids.